Migrated `pkg/controller/endpointslicemirroring` to contextual logging

Namanl2001 opened the pull request:
Signed-off-by: Naman <namanlakhwani@gmail.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
Use contextual logging for `pkg/controller/endpointslicemirroring`
#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Part of https://github.com/kubernetes/enhancements/issues/3077

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Migrated the `EndpointSlice Mirroring` controller (within `kube-controller-manager`) to use [contextual logging](https://k8s.io/docs/concepts/cluster-administration/system-logs/#contextual-logging).
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/3077-contextual-logging
```

/wg structured-logging
/area logging
/assign @pohly

Namanl2001 committed:
migrated pkg/controller/endpointslicemirroring to contextual logging

Signed-off-by: Naman <namanlakhwani@gmail.com>

Namanl2001 commented:
/remove-sig api-machinery
/remove-sig apps
/remove-sig testing
/remove-sig network

/sig instrumentation
/priority important-longterm

yangjunmyfm192085 commented on pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go:
```diff
@@ -280,23 +283,23 @@ func (c *Controller) syncEndpoints(key string) error {
 	if err != nil {
 		if apierrors.IsNotFound(err) {
-			klog.V(4).Infof("%s/%s Endpoints not found, cleaning up any mirrored EndpointSlices", namespace, name)
+			logger.V(4).Info("Endpoints not found, cleaning up any mirrored EndpointSlices", "namespace", namespace, "name", name)
```
```suggestion
			logger.V(4).Info("Endpoints not found, cleaning up any mirrored EndpointSlices", "endpoints", klog.KRef(namespace, name))
```

yangjunmyfm192085 commented on pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go:
```diff
@@ -280,23 +283,23 @@ func (c *Controller) syncEndpoints(key string) error {
 
 	if !c.shouldMirror(endpoints) {
-		klog.V(4).Infof("%s/%s Endpoints should not be mirrored, cleaning up any mirrored EndpointSlices", namespace, name)
+		logger.V(4).Info("Endpoints should not be mirrored, cleaning up any mirrored EndpointSlices", "namespace", namespace, "name", name)
```
```suggestion
		logger.V(4).Info("Endpoints should not be mirrored, cleaning up any mirrored EndpointSlices", "endpoints", klog.KRef(namespace, name))
```

yangjunmyfm192085 commented on pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go:
```diff
@@ -280,23 +283,23 @@ func (c *Controller) syncEndpoints(key string) error {
 	if err != nil {
 		if apierrors.IsNotFound(err) {
-			klog.V(4).Infof("%s/%s Service not found, cleaning up any mirrored EndpointSlices", namespace, name)
+			logger.V(4).Info("Service not found, cleaning up any mirrored EndpointSlices", "namespace", namespace, "name", name)
```
```suggestion
			logger.V(4).Info("Service not found, cleaning up any mirrored EndpointSlices", "service", klog.KRef(namespace, name))
```

yangjunmyfm192085 commented on pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go:
```diff
@@ -305,7 +308,7 @@ func (c *Controller) syncEndpoints(key string) error {
 	// If a selector is specified, clean up any mirrored slices.
 	if svc.Spec.Selector != nil {
-		klog.V(4).Infof("%s/%s Service now has selector, cleaning up any mirrored EndpointSlices", namespace, name)
+		logger.V(4).Info("Service now has selector, cleaning up any mirrored EndpointSlices", "namespace", namespace, "name", name)
```
```suggestion
		logger.V(4).Info("Service now has selector, cleaning up any mirrored EndpointSlices", "service", klog.KRef(namespace, name))
```

yangjunmyfm192085 commented on pkg/controller/endpointslicemirroring/reconciler.go:
```diff
@@ -88,7 +88,7 @@ func (r *reconciler) reconcile(endpoints *corev1.Endpoints, existingSlices []*di
 			} else {
 				numInvalidAddresses++
-				klog.Warningf("Address in %s/%s Endpoints is not a valid IP, it will not be mirrored to an EndpointSlice: %s", endpoints.Namespace, endpoints.Name, address.IP)
+				logger.Info("Address in Endpoints is not a valid IP, it will not be mirrored to an EndpointSlice.", "namespace", endpoints.Namespace, "name", endpoints.Name, "IPaddress", address.IP)
```
```suggestion
				logger.Info("Address in Endpoints is not a valid IP, it will not be mirrored to an EndpointSlice", "endpoints", klog.KObj(endpoints), "IP", address.IP)
```

yangjunmyfm192085 commented on pkg/controller/endpointslicemirroring/reconciler.go:
```diff
@@ -103,7 +103,7 @@ func (r *reconciler) reconcile(endpoints *corev1.Endpoints, existingSlices []*di
 			} else {
 				numInvalidAddresses++
-				klog.Warningf("Address in %s/%s Endpoints is not a valid IP, it will not be mirrored to an EndpointSlice: %s", endpoints.Namespace, endpoints.Name, address.IP)
+				logger.Info("Address in Endpoints is not a valid IP, it will not be mirrored to an EndpointSlice.", "namespace", endpoints.Namespace, "name", endpoints.Name, "IPaddress", address.IP)
```
```suggestion
				logger.Info("Address in Endpoints is not a valid IP, it will not be mirrored to an EndpointSlice", "endpoints", klog.KObj(endpoints), "IP", address.IP)
```

yangjunmyfm192085 commented on pkg/controller/endpointslicemirroring/reconciler.go:
```diff
@@ -124,7 +124,7 @@ func (r *reconciler) reconcile(endpoints *corev1.Endpoints, existingSlices []*di
 	// addresses exceeding MaxEndpointsPerSubset.
 	if addressesSkipped > numInvalidAddresses {
-		klog.Warningf("%d addresses in %s/%s Endpoints were skipped due to exceeding MaxEndpointsPerSubset", addressesSkipped, endpoints.Namespace, endpoints.Name)
+		logger.Info("A few addresses in Endpoints were skipped due to exceeding MaxEndpointsPerSubset", "skipped", addressesSkipped, "namespace", endpoints.Namespace, "name", endpoints.Name)
```
```suggestion
		logger.Info("Addresses in Endpoints were skipped due to exceeding MaxEndpointsPerSubset", "skippedAddresses", addressesSkipped, "endpoints", klog.KObj(endpoints))
```

yangjunmyfm192085 commented:
For log structure, please refer to the documentation
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md

Namanl2001 commented:
> For log structure, please refer to the documentation https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/migration-to-structured-logging.md

I will go through this doc again 😅 
Thanks for the detailed review :) addressed all the comments.

logicalhan commented:
/triage accepted

cici37 commented:
/remove-sig api-machinery

Namanl2001 commented:
@pohly Can I have a `/lgtm` label on this PR as well?

pohly commented on pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go:
```diff
@@ -239,38 +241,39 @@ func (c *Controller) processNextWorkItem() bool {
 
-	err := c.syncEndpoints(cKey.(string))
-	c.handleErr(err, cKey)
+	logger := klog.FromContext(context.TODO())
```
Let's avoid *adding* TODO items as part of these changes - we want to *remove* them :grin: 

In this case the solution used elsewhere (check pending PRs) is to add a `logger` parameter pass that through from `Run`.

Namanl2001 replied:
this similar case is with another component and I added a comment [here](https://github.com/kubernetes/kubernetes/pull/115295#discussion_r1110193847)

pohly replied:
See https://github.com/kubernetes/kubernetes/pull/115295#discussion_r1132343721

Namanl2001 commented on pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go:
```diff
@@ -239,38 +241,39 @@ func (c *Controller) processNextWorkItem() bool {
 	}
 
-	klog.Warningf("Retry budget exceeded, dropping %q Endpoints out of the queue: %v", key, err)
+	logger.Info("Retry budget exceeded, dropping Endpoints out of the queue", "endpoints", key, "err", err)
```
addressed, Please check if [this](https://github.com/kubernetes/kubernetes/pull/114982/commits/bc3cfc5c49e92ae9e433c8f834f60a0707a25d28) looks good 🙏

pohly commented on pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go:
```diff
@@ -254,31 +254,36 @@ func (c *Controller) handleErr(logger klog.Logger, err error, key interface{}) {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
```
This changes the function logic: `metrics.EndpointsSyncDuration.WithLabelValues().Observe(syncDuration)` will not be called anymore when this error occurs.

Better leave this as it was and log `"key", key`.

pohly commented on pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go:
```diff
@@ -254,31 +254,36 @@ func (c *Controller) handleErr(logger klog.Logger, err error, key interface{}) {
+	if err != nil {
+		logger.Info("Error splitting key into name and namespace")
+		return
+	}
```
This changes the function logic. It's a small change, but I prefer to not modify the control flow just to make logs better.

Can you shift this back and log `"key", key` in the next log entry?

pohly commented on pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go:
```diff
@@ -208,69 +209,70 @@ func (c *Controller) Run(workers int, stopCh <-chan struct{}) {
 	}()
 
-	klog.V(4).Infof("syncEndpoints(%q)", key)
+	logger.V(4).Info("syncEndpoints", "endpoints", key)
```
"endpoints" is still not used consistently.

pohly commented on test/integration/endpointslice/endpointslicemirroring_test.go:
```diff
@@ -80,7 +80,7 @@ func TestEndpointSliceMirroring(t *testing.T) {
 	go epController.Run(ctx, 5)
 	go epsController.Run(5, ctx.Done())
-	go epsmController.Run(5, ctx.Done())
+	go epsmController.Run(ctx, 5)
```
While at it, convert these tests to per-test output: create a `_, ctx := ktesting.NewTestContext(t)` for each test and use that context instead of `context.Background()`.

pohly reviewed (commented):
As part of your next push, please squash without rebasing.

Also please remove the exception from https://github.com/kubernetes/kubernetes/blob/5f23f835879b514e9bf16809059b38957d20a486/hack/logcheck.conf#L47

Namanl2001 commented:
> As part of your next push, please squash without rebasing.
> 
> Also please remove the exception from
> 
> https://github.com/kubernetes/kubernetes/blob/5f23f835879b514e9bf16809059b38957d20a486/hack/logcheck.conf#L47

The exceptions are not present, will have to reabse to remove it as a part of this PR

pohly commented on test/integration/endpointslice/endpointslicemirroring_test.go:
```diff
@@ -75,12 +76,11 @@ func TestEndpointSliceMirroring(t *testing.T) {
 	// Start informer and controllers
-	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
+	_, ctx := ktesting.NewTestContext(t)
```
The context from `NewTestContext` does not get canceled. You have to keep the `WithCancel` here to avoid the goroutine leak from https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/114982/pull-kubernetes-integration/1673310069199998976.

pohly commented on test/integration/endpointslice/endpointslicemirroring_test.go:
```diff
@@ -330,10 +330,9 @@ func TestEndpointSliceMirroringUpdates(t *testing.T) {
 	// Start informer and controllers
-	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
+	_, ctx := ktesting.NewTestContext(t)
```
Same here.

Namanl2001 committed:
added withCancel()

Signed-off-by: Naman <namanlakhwani@gmail.com>

pohly commented on test/integration/endpointslice/endpointslicemirroring_test.go:
```diff
@@ -505,10 +508,9 @@ func TestEndpointSliceMirroringSelectorTransition(t *testing.T) {
-	defer close(stopCh)
-	informers.Start(stopCh)
-	go epsmController.Run(1, stopCh)
+	_, ctx := ktesting.NewTestContext(t)
```
Another missing `ctx, cancel := context.WithCancel(ctx)` here.

Namanl2001 committed:
add withCancel

Signed-off-by: Naman <namanlakhwani@gmail.com>

Namanl2001 commented:
Changes of this PR were cherry picked and are merged with #115295
